KOGITO-7999: Stunner - Domain / bundle customization in js

handreyrc opened the pull request:
Hey @romartin, could you please review this PR?  

**JIRA**: [KOGITO-7999](https://issues.redhat.com/browse/KOGITO-7999)

Thanks!

romartin commented on packages/serverless-workflow-diagram-editor/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/definition/DefinitionSetResourceType.java:
```diff
@@ -18,6 +18,7 @@
 import org.uberfire.workbench.type.ResourceTypeDefinition;
 
+// TODO: Since kogito, this can be removed.
 public interface DefinitionSetResourceType extends ResourceTypeDefinition {
```
can you please @handreyrc try to clean this up? Also related resources, if necessary (eg: top class)

romartin commented on packages/serverless-workflow-diagram-editor/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/definition/jsadapter/JsMorphAdapter.java:
```diff
@@ -0,0 +1,60 @@
+import org.kie.workbench.common.stunner.core.definition.morph.MorphDefinition;
+import org.kie.workbench.common.stunner.core.definition.morph.MorphProperty;
+
+// TODO
```
I would drop the  TODO, for now, and maybe just rename the class to _EmptyJsMorhAdapter_

romartin commented on packages/serverless-workflow-diagram-editor/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/api/JsStunnerSession.java:
```diff
@@ -0,0 +1,113 @@
+        return definition;
+    }
+
+    // TODO: Multiple selection / center selection.
```
remove the TODO

romartin commented on packages/serverless-workflow-diagram-editor/lienzo-core/src/main/java/com/ait/lienzo/client/core/types/JsCanvas.java:
```diff
@@ -88,12 +92,58 @@ public Viewport getViewport() {
     }
 
+    public void translate(double x, double y) {
+        // TODO: Composite transform ops or create new matrix?
```
Test this method from the UI, and it works, remove the TODO

romartin replied:
maybe it is clear to create a new transform instance every time this method is called

romartin commented on packages/serverless-workflow-diagram-editor/lienzo-core/src/main/java/com/ait/lienzo/client/core/types/JsCanvas.java:
```diff
@@ -88,12 +92,58 @@ public Viewport getViewport() {
+    }
+
+    public void scale(int factor) {
+        // TODO: Composite transform ops or create new matrix?
```
Same thing as transform above.

romartin commented on packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/Workflow.java:
```diff
@@ -46,92 +35,52 @@
     @JsonbTypeSerializer(EventJsonbTypeSerializer.class)
     @JsonbTypeDeserializer(EventJsonbTypeDeserializer.class)
-    private Object events; //TODO array or string
+    public Object events; //TODO array or string
```
drop this TODO comment

romartin commented on packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/README.md:
```diff
@@ -89,8 +90,51 @@ Navigate to the test page in your root context path, eg: `http://localhost:8080/
         var s = jsl.getWiresShape('_A9481DBC-3E87-40EE-9925-733B24404BC0')
         s.getChild(1).fillColor = "red"
+
+# TODO: FOR DEMO
```
Drop this TODO 
The goal is to describe the public API for jS (JsStunnerEditor and sbuclasses, aslo the JsCanvas)
Also show some examples.

handreyrc commented:
Hey @romartin, we are good for a 2nd round of reviews here!

Thanks!

romartin commented on packages/serverless-workflow-diagram-editor/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-api/kie-wb-common-stunner-core-api/src/main/java/org/kie/workbench/common/stunner/core/factory/definition/JsDefinitionFactory.java:
```diff
@@ -14,17 +14,24 @@
+        return true;
+    }
+
+    private native Object createInstanceForType(String typeName) /*-{
```
is it possible to do this in j2cl and so drop the use of jsni?

handreyrc replied:
Hey @romartin,

I spent sometime looking into it when I was working here, however, till the moment I couldn't find any documented alternative that do not expose this method or its internals to the world.
IMO, for security reasons we should keep it private.

WDYT?

romartin commented on packages/serverless-workflow-diagram-editor/package.json:
```diff
@@ -21,8 +21,8 @@
     "build:prod": "run-script-os",
-    "build:prod:linux:darwin": "pnpm lint && mvn clean install -DskipTests=$(build-env tests.run --not) -Dmaven.test.failure.ignore=$(build-env tests.ignoreFailures) -DskipITs=$(build-env integrationTests.run --not) -Pno-showcase",
-    "build:prod:win32": "pnpm lint && pnpm powershell \"mvn clean install `-DskipTests=$(build-env tests.run --not) `-Dmaven.test.failure.ignore=$(build-env tests.ignoreFailures) `-DskipITs=$(build-env integrationTests.run --not) `-Pno-showcase\"",
+    "build:prod:linux:darwin": "pnpm lint && mvn clean install -DskipTests -Dmaven.test.failure.ignore=$(build-env tests.ignoreFailures) -DskipITs -Pno-showcase",
```
is really necessary to change this? are we skipping some tests?

handreyrc replied:
Good catch @romartin!
I should have put it back at the end.

Thanks!

romartin commented on packages/serverless-workflow-diagram-editor/package.json:
```diff
@@ -21,8 +21,8 @@
-    "build:prod:linux:darwin": "pnpm lint && mvn clean install -DskipTests=$(build-env tests.run --not) -Dmaven.test.failure.ignore=$(build-env tests.ignoreFailures) -DskipITs=$(build-env integrationTests.run --not) -Pno-showcase",
-    "build:prod:win32": "pnpm lint && pnpm powershell \"mvn clean install `-DskipTests=$(build-env tests.run --not) `-Dmaven.test.failure.ignore=$(build-env tests.ignoreFailures) `-DskipITs=$(build-env integrationTests.run --not) `-Pno-showcase\"",
+    "build:prod:linux:darwin": "pnpm lint && mvn clean install -DskipTests -Dmaven.test.failure.ignore=$(build-env tests.ignoreFailures) -DskipITs -Pno-showcase",
+    "build:prod:win32": "pnpm lint && pnpm powershell \"mvn clean install `-DskipTests `-Dmaven.test.failure.ignore=$(build-env tests.ignoreFailures) ` -DskipITs` -Pno-showcase\"",
```
is really necessary to change this? are we skipping some tests?

romartin reviewed (approved):
Hey Handrey!

Code looks good to me!!! I just added a few (not blocking) review comments, try to fix if itt possible please. Anyway, that's a great job, thanks!!!

@LuboTerifaj **feel free to start the review**. This PR is just about "internal" updates on the code side, but from the end user perspective, nothing really should change. 
So just feel free to do some basic smoketesting and we can also rely on the integration tests during the build. 
Anyway let's please try to move this forward and merge it asap, it has many changes and it provides some out-of-the-box benefits, such as a few less memory consumption, as we're dropping several generated classes and objects.

Thanks!

romartin commented:
@LuboTerifaj all fixed I guess. Ready for your review! See my previous comment https://github.com/kiegroup/kie-tools/pull/1366#pullrequestreview-1260741313

Thanks!

romartin commented on packages/serverless-workflow-diagram-editor/kie-wb-common-stunner/kie-wb-common-stunner-core/kie-wb-common-stunner-commons/kie-wb-common-stunner-client-common/src/main/java/org/kie/workbench/common/stunner/core/client/api/JsStunnerSession.java:
```diff
@@ -0,0 +1,112 @@
+
+    public Object getSelectedDefinition() {
+        Node node = getSelectedNode();
+        View<?> content = (View<?>) node.getContent();
```
Hey @handreyrc 

I just noticed right now, once testing this API from js, that if there is no element selected in the canvas, this method throws an exception, also in js. IMO should be better to return `undefined` or `null`, in case nothing selected, so we don't drive js developers to constantly use `try/catch` blocks.

This issue may also happen with other methods in this class, for example getSelectedNode, getSelectedEdge, etc. 

Can you please try to fix those methods in order not to throw exceptions when nothing is selected? This way, once merging this., if someone starts using this API, we'll make it easier to work with. WDYT?

Thanks!

handreyrc replied:
Hey @romartin,
I agree with you. I reviewed the JsClasses and there aren't many.
It's already fixed!
Thanks

romartin replied:
Great, thanks so much!

LuboTerifaj reviewed (commented):
Hello @handreyrc , @romartin ,

It seems the API works pretty good, but there are several points that need to be clarified/updated.

I tested all functions listed in jira from Canvas API and Session API, and Domain Definitions API unless specified otherwise in my comments below.
Regarding the following functions, you can find my concerns, comments, or found issues next to each function.

#### Canvas API

- `setBackgroundColor(id: string, color: string)`
  - It is not clear what format should a user use to set up the color properly. What are the allowed values? `// String, Hex, RGB, etc. ("green", "magenta", "#ffff66", [255, 255, 0])`?
  - Default color is black. I think it should be white.
		
- `getBorderColor(id: string)`
  - Function returns null.

- `setBorderColor(id: string)`
  - Function doesn't work.

- `center(id: string)`
  - What does it mean center? I expect to center the node, but it is not clear, what is it supposed to be relative to.
  - I wasn't able to see any results of the function.
	
- `calculateCenter(id: string)`
  - It is not clear what the returned coordinates are relative to. There are more possible expectations from users: center of canvas; X0, Y0 of canvas; respectively something different.
	
- `translate(x: Number, y: Number)`
  - The function name is not very intuitive. Would it make sense to change it? Is there any specific reason why it should stay like it is?
	
- `rotateGroupOverCenter(group: Group, degrees: Number, duration: Number)`
  - There is missing the description of what is Group parameter and how to use it.
  -  See also **other thoughts** section below
	
- `getNativeContext()`
  - Function doesn't work.
  - There is only getNativeContent() function.
	
- `draw()`
  - It should be properly described, when it is necessary to call this function.
	
- `add(shape: Primitive)`
  - There is missing info about primitive shape, its required fields and how to use the function properly.
  -  See also **other thoughts** section below
	
- `getWiresManager()`
  - There is missing information how it can be used.

- `log()`
  - There is missing information how it can be used.
	
- `getNodeIdSet()`
  - Why there is this function if we have getNodeIds()? Is it on purpose?
	
- `getWiresShape(id: string)`
  - What is it used for? It's not clearly described.
  - I was not able to use it properly.
	
#### Session API

- `getEdgeByUUID(id: string)`
  - There is nothing wrong with the function itself, but there is no easy way how to find the edge UUID. There is no function to get all edges and their UUIDs, but only selected element UUID. The only way how to get Edge is to select the edge first.
	
- `getSelectedDefinition()`
  - It seems it works as expected, but it is not described properly what the definition is and how it can be used.
	
#### Domain Definitions API

There is missing information how it can be used.
There is no function for getting the objects, or described how to create the objects, for the following parameters from API functions:
- Object definitionSet
- Annotation domainQualifier
- RuleSet ruleSet

See also **other thoughts** section below.

The functions don't log anything and don't have any return values.

As a result, I couldn't test initialization functions properly.

#### Other thoughts

What all functions are supposed to be accessible? Do we want expose all of them?	
For example, the functions that are not part of the API:	
- `contentWindow.com.ait.lienzo.client.core.shape.Group()`
- `org.kie.workbench.common.stunner.sw.definition.InjectState()`
	
I think it would be beneficial to have better documentation for the functions behaviour and how to properly access them via various streams. The code documentation would help to achieve this more easily (see my next point).

Regarding code, I think it would worth adding Javadoc for the classes and methods, but I understand it is not a standard for this repo.
Please see also my inline comments.

We need to add also some integration tests, but I understand it is probably out of scope of this jira. I might help with that in a separate jira.

Thank you for this PR! It is really great to have the API that can be also used for testing.

LuboTerifaj commented:
Hey @caponetto , this PR is already reviewed and it is not waiting for qe at the current stage. Removing waiting-for-qe label.

caponetto commented:
@LuboTerifaj I would keep this label until QE approves the PR.

LuboTerifaj commented:
@caponetto I see. I thought it means the PR was not reviewed by QE yet. The label name is not straight forward. Are the the labels used documented anywhere? Thanks!

caponetto commented:
@LuboTerifaj not that I know of. But, well, this PR is still waiting for QE as QE hasn't approved it yet.
That's my understanding.

LuboTerifaj commented:
Hey @caponetto ,

It depends on point of view. One could interpret it like this: since QE provided review with comment or rejection, it means it is not approved, but no work is required from QE at the current stage, so we are not waiting for qe.

Therefore I was confused. Something like `waiting-for-qe-approval` would make more sense here, although it might be too long label. The other option would be to document it, but I think the self descriptive name is a better option.

Anyway, now I know what is probably meant by that.

Thanks!

handreyrc commented on packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-api/src/main/java/org/kie/workbench/common/stunner/sw/definition/CallbackState.java:
```diff
@@ -17,34 +17,15 @@
-import org.kie.workbench.common.stunner.core.definition.annotation.morph.Morph;
 
-/**
- * The Callback state allows you to explicitly model manual decision steps during workflow execution.
```
Hey @LuboTerifaj,

Those comments in the model were just for reference as we learned about the specification. 
We are not maintaining the comments up-to-date with the specification so in my opinion it is better to remove them.

WDYT?

handreyrc commented on packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/README.md:
```diff
@@ -73,24 +73,87 @@ Navigate to the test page in your root context path, eg: `http://localhost:8080/
+  | scaleWithXY           | Number x, <br/>Number y                                |                   | scale the diagram (zoom in/out)                          |
+  | getScaleX             |                                                        |      Number       | get X scale factor                                       |
+  | getScaleY             |                                                        |      Number       | get Y scale factor                                       |
+  | rotateGroupOverCenter | Group group, <br/>Number degrees, <br/>Number duration |      Number       | get Y scale factor                                       |
```
@LuboTerifaj  I removed the method "rotateGroupOverCenter".
Actually it was part of a demo do show what can be done using the API so it should be removed.

handreyrc commented on packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/README.md:
```diff
@@ -73,24 +73,87 @@ Navigate to the test page in your root context path, eg: `http://localhost:8080/
+  | rotateGroupOverCenter | Group group, <br/>Number degrees, <br/>Number duration |      Number       | get Y scale factor                                       |
+  | getLayer              |                                                        |       Layer       | get HTMLCanvasElement element                            |
+  | getCanvas             |                                                        | HTMLCanvasElement | get canvas Layer instance                                |
+  | getViewport           |                                                        |     Viewport      | get canvas NativeContext2D instance                      |
```
@LuboTerifaj ,
The description of these two methods are inverted.
It's fixed now.

handreyrc commented on packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/README.md:
```diff
@@ -73,24 +73,87 @@ Navigate to the test page in your root context path, eg: `http://localhost:8080/
+  | getLayer              |                                                        |       Layer       | get HTMLCanvasElement element                            |
+  | getCanvas             |                                                        | HTMLCanvasElement | get canvas Layer instance                                |
+  | getViewport           |                                                        |     Viewport      | get canvas NativeContext2D instance                      |
+  | getNativeContext      |                                                        |  NativeContext2D  | get canvas Viewport instance                             |
```
@LuboTerifaj ,
The description of these two methods are inverted.
It's fixed now.

handreyrc commented:
Hey @LuboTerifaj and @romartin 

I went through the comments above, please check my comments below.

 #### Canvas API
 * `setBackgroundColor(id: string, color: string)`
   
   * It is not clear what format should a user use to set up the color properly. What are the allowed values? `// String, Hex, RGB, etc. ("green", "magenta", "#ffff66", [255, 255, 0])`? 
       * The function description was updated to help with the supported color formats.
   * Default color is black. I think it should be white. 
      * @LuboTerifaj, how did you get to default value?
 * `getBorderColor(id: string)`
   
   * Function returns null.
      * Fixed 
 * `setBorderColor(id: string)`
   
   * Function doesn't work.
      * Fixed
 * `center(id: string)`
   
   * What does it mean center? I expect to center the node, but it is not clear, what is it supposed to be relative to.
   * I wasn't able to see any results of the function.
      * The description of the function was updated. This function is applicaple only to nodes out of the viewable area of the canvas. To test it load a large workflow pick a node(uuid) in the edge of the diagram, then increase the zoom until the node is not visible anymore, finally call the center function for the non visible node.
 * `calculateCenter(id: string)`
   
   * It is not clear what the returned coordinates are relative to. There are more possible expectations from users: center of canvas; X0, Y0 of canvas; respectively something different.
     * The description of the function was updated. This function is applicaple to nodes out of the viewable area of the canvas. It gives the values to be used in the "translate" function to center a non visible node in the canvas.
 * `translate(x: Number, y: Number)`
   
   * The function name is not very intuitive. Would it make sense to change it? Is there any specific reason why it should stay like it is?
     * Unfortunately in my opinion we should not change the name of this function. It is part of the Viewport API (viewort>transform>translate) so we can make developers confuse if we change it. 
 * `rotateGroupOverCenter(group: Group, degrees: Number, duration: Number)`
   
   * There is missing the description of what is Group parameter and how to use it.
      * I removed this method, it was used to showcase the API possibilities in the past and should be removed. 

 * `getNativeContext()`
   
   * Function doesn't work.
   * There is only getNativeContent() function.
      * Fixed. The name of the function 'getNativeContent()' was wrong it should be 'getNativeContext()'. It works, it returns the NativeContext2D instance. A low level API to render 2D graphics.
 * `draw()`
   
   * It should be properly described, when it is necessary to call this function.
      * The description of the function was updated. It should be called everytime the lookandfeel is changed.
 * `add(shape: Primitive)`
   
   * There is missing info about primitive shape, its required fields and how to use the function properly.
      * This function allows to add a new primitive shape to the canvas. In lienzo there is the concept of primitives/ shapes / nodes /wiresShapes. I think it would be a bit complicated to cover those implementation concepts in the function description.
      * e.g.  
         *  var prim = new com.ait.lienzo.client.core.shape.Rectangle(50, 50);
         *  prim.x = 100; prim.y = 100;
         * canvas.add(prim); canvas.draw();
   * See also **other thoughts** section below
 * `getWiresManager()`
   
   * There is missing information how it can be used.
      * The function description was updated. This is another low level API, it acts upon the wiresShapes by grouping and extending the capabilities on top of the primitive shapes.
 * `log()`
   
   * There is missing information how it can be used.
      * The function description was updated. The idea is to provide a way to recover information about the diagram. There is still a lot of room to improvements here for now check the example below. It shows the how the nodes(uuids) are connected between them.
      * e.g.  canvas.log().logWiresShapes()
   
 * `getNodeIdSet()`
   
   * Why there is this function if we have getNodeIds()? Is it on purpose?
      * It was on purpose, they return diferent types. getNodeIds() return a NFastArrayList<String> and getNodeIdSet() returns a Set<String> that is needed to recover flatShapes(shapes in a group) in the JsWiresShape class.
 * `getWiresShape(id: string)`
   
   * What is it used for? It's not clearly described.
   * I was not able to use it properly.
      * When the functions set/getBackground color is called we are dealing with wiresShapes under the hood. I you want to have access to all the wiresShape API (JsWiresShape) you can pick a WiresShape instance and use the methods exposed. 
      * e.g. canvas.setBackgroundColor('uuid' , 'color')  is a shortcut for  canvas.getWiresShape('uuid').setBackgroundColor('green')
 
 #### Session API
 * `getEdgeByUUID(id: string)`
   
   * There is nothing wrong with the function itself, but there is no easy way how to find the edge UUID. There is no function to get all edges and their UUIDs, but only selected element UUID. The only way how to get Edge is to select the edge first.
     * As  far as I know getting the edges uuids is a costly operation because we have to traverse the graph to get those. However, it is something we could consider for the next iterations and improvements in the API.
 * `getSelectedDefinition()`
   
   * It seems it works as expected, but it is not described properly what the definition is and how it can be used.
 
 #### Domain Definitions and Domain API and related
The APIs below are in the initial stage of development and are prety basic. They are aimed part of the configuration of the Domain, Definitions and rules via js. They are functional, however, in my opinion they should be evaluated as part of @romartin work on top of this PR so we could have the whole picture. 
WDYT? 

 * `getSelectedDefinition()`
 * Object definitionSet
 * Annotation domainQualifier
 * RuleSet ruleSet
 
 
 #### Other thoughts
 What all functions are supposed to be accessible? Do we want expose all of them? For example, the functions that are not part of the API:
 
 * `contentWindow.com.ait.lienzo.client.core.shape.Group()`
 * `org.kie.workbench.common.stunner.sw.definition.InjectState()`
     * Classes anotated with @JsType can be exposed if their methods are public. It means they are not ofuscated when transpiled to js. If there are methods available in the APIs (Editor, Canvas, Session, Domain...) to get to these objects, they can be consumed. 
 
 #### In regards with the documentation
@LuboTerifaj definetely your concerns about documentation are justified.
In my opinion we have somehow to find a beter way to share this knowledge other than doing it trough the code. API documentation often is restricted to the scope of each function exposed. When it comes on how to put things together maybe a tutorial alike format is more suitable.  Maybe it is something we may keep in mind for the near future thinking of Stunner as a Framewok.

WDYT guys?

handreyrc commented:
Hey @romartin and @LuboTerifaj,

We are still working on an external unrelated issue to get this PR green.
In the meantime  the vsix plugin below can be used for reviewing the latest.

**VS Code**: [plugin](https://drive.google.com/file/d/16Il6g6nLWnVKwS7V9Kg9JsgDquVAMLsz/view?usp=sharing)

Thanks!

handreyrc commented:
Hey @romartin and @LuboTerifaj,

Once the PR https://github.com/kiegroup/kie-tools/pull/1536 was merged, this PR is up-to-date, green and ready for anorther round of reviews.

Thanks!

LuboTerifaj reviewed (changes requested):
Hello @handreyrc ,

See my comments below. They are based on my previous [review](https://github.com/kiegroup/kie-tools/pull/1366#pullrequestreview-1289064876) and your [comment](https://github.com/kiegroup/kie-tools/pull/1366#issuecomment-1472996513) on it.

### Canvas API

When you set node color using function: `canvas.setBackgroundColor(nodeId, "default")`, the black color will be set.
- It should be the white.

`center(id: string)`
- I am still not able to use the function. It doesn't matter how many nodes the diagram contains, all of them are always visible. The center function didn't work even when I zoomed in, in a way that not all nodes where visible.

`calculateCenter(id: string)`
- It is still not clear what the returned coordinates are relative to. This needs to be clarified.
- What can be this function used for?
- The function may return non integer values. Is this expected?

`add(shape: Primitive)`
- The function works properly.
- If we are exposing API and expecting the specific parameter, it should be described properly where such parameter can be taken.

`getWiresManager()`
- Still missing the information how it can be used.

`log()`
- I see there is one example `canvas.log().logWiresShapes()`. Are there any other functions that can be used there? If yes, this should be properly documented.

`getNodeIdSet()`
- I understand it returns a different type, but I don't understand why it is needed. Could you please clarify it to me?

`getWiresShape(id: string)`
- Missing documentation
- In addition, what functions can be used from WiresShape instance?

### Session API

getEdgeByUUID(id: string)
- I am okay with **reporting this** in separate jira.

`getSelectedDefinition()`
- I didn't find any changes in comparison from my previous review.

### Domain Definitions API

If we want to implement it in separate jira, it's okay. Just please remove all related code from the PR and update the jira's description.

Based on your comment, I didn't retest this API. Please let me know what is the plan with that, so I know if it needs to be retested or not.

### Other thoughts

I see that "_Classes annotated with @jstype can be exposed if their methods are public_". Do you know if it is possible to somehow avoid such behaviour? I mean to hide undesired functions?

If we don't change anything, I think we should document at least those that are needed to use in the API we are exposing in Canvas, Session and Domain. I mean to basically document properly the exposed API.


I agree, we should create the more complex documentation with real examples and it should be the following up task.
However, I think the basic usage should be documented properly as it is usually done for any API.

Thank you!

handreyrc commented:
Hey @LuboTerifaj and @romartin ,

I went though the latest comments above and I made some changes.
After analyzing the the methods I  reached the conclusion that some methods we were exposing are helpers to other methods in the API. So I refactored the code to remove these helpers because they don't make sense from the point of view of the user of the API and can be confusing.

 ### Canvas API
 When you set node color using function: `canvas.setBackgroundColor(nodeId, "default")`, the black color will be set.
 
 * It should be the white.
    * There is something else happening here. Actually 'default' or any unsuported color name is accepted and set to the shape internally. However, if it is not a valid color it is somehow set to black. The curious it that if you try to recover the current bacgroundcolor after setting an invalid color the result is not 'black' or '#000000', it is the string informed.  Maybe e need to add some sort of validatation to lienzo-core to avoid this incosistency. I suggest  to open a separated Jira for this fix.
 
 `center(id: string)`
 
 * I am still not able to use the function. It doesn't matter how many nodes the diagram contains, all of them are always visible. The center function didn't work even when I zoomed in, in a way that not all nodes where visible.
    * Lets sync about the usage of this method.
    
 `calculateCenter(id: string)`
 
 * It is still not clear what the returned coordinates are relative to. This needs to be clarified.
 * What can be this function used for?
 * The function may return non integer values. Is this expected?
    * This method is a helper to the center method so it was removed from the API it does not make sense to expose it to the user. 
 
 `add(shape: Primitive)`
 
 * The function works properly.
 * If we are exposing API and expecting the specific parameter, it should be described properly where such parameter can be taken.
    * More information on how to get to the primitive types was added to the documentation. 
 
 `getWiresManager()`
 
 * Still missing the information how it can be used.
    * This is helper method to other methods and the WiresManager object is not exposed(@JsType) so this method was removed. 
 
 `log()`
 
 * I see there is one example `canvas.log().logWiresShapes()`. Are there any other functions that can be used there? If yes, this should be properly documented.
    * I talked about this method with @romartin and we agree that it may not make sense to the user. Users should create their own logs in a way they make sense to them so this method was also removed from the API. 
 
 `getNodeIdSet()`
 
 * I understand it returns a different type, but I don't understand why it is needed. Could you please clarify it to me?
    * I agree, it is redundant to have two methods returning the same data. So the 'getNodeIdSet' mathod was removed and the code refactored to use the getNodeIds instead.  
 
 `getWiresShape(id: string)`
 
 * Missing documentation
 * In addition, what functions can be used from WiresShape instance?
    * More information was added to the documentation. Basically it has everthing a shape has, but also grouping and connection capabilities through the magnets. 
 
 ### Session API
 getEdgeByUUID(id: string)
 
 * I am okay with **reporting this** in separate jira.
    * I agree it is nice fearture to have. 
 
 `getSelectedDefinition()`
 
 * I didn't find any changes in comparison from my previous review.
    * My bad I missed this one. It gets the model object for the selected node then it is possible to recover those properties values. 
 
 ### Domain Definitions API
 If we want to implement it in separate jira, it's okay. Just please remove all related code from the PR and update the jira's description.
 
 Based on your comment, I didn't retest this API. Please let me know what is the plan with that, so I know if it needs to be retested or not.
 * Lets sync with @romartin about the best approach here. 
 
 ### Other thoughts
 I see that "_Classes annotated with @jstype can be exposed if their methods are public_". Do you know if it is possible to somehow avoid such behaviour? I mean to hide undesired functions?
* Actually it is not possible to hide code. When the code is transpiled from java to js we lose the access modifiers (private public). In this case the best we can do is to anotate the methods we don't want to be used with @JsIgnore. It will cause to traspiler to obfuscate them, however, they will still be there anyway. In a rough way, exposing a method means it is not obfuscated and can be easily read by the user.

Please lets sync @romartin and @LuboTerifaj if we still need more work here.

Thanks!

romartin commented on packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/README.md:
```diff
@@ -73,24 +73,60 @@ Navigate to the test page in your root context path, eg: `http://localhost:8080/
+  | clearSelection         |                   |             | clear selection                                                                                                            |
+  | getSelectedNode        |                   |    Node     | get selected Node instance                                                                                                 |
+  | getSelectedEdge        |                   |    Edge     | get selected Edge instance                                                                                                 |
+  | getSelectedDefinition  |                   | Definition  | get selected Definition instance <br/> it is the data model of the selelected node with all declared properties and values |
```
Hey @handreyrc 
I suggest to change the return type from `Definition` to just an `Object`. I results more usual for third parties, when dealing with JS API's.... WDYT?

handreyrc replied:
Hey @romartin! 

I was looking for a better way to describe it but I was stuck on java side. 

Thanks!

LuboTerifaj commented on packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/README.md:
```diff
@@ -73,24 +73,60 @@ Navigate to the test page in your root context path, eg: `http://localhost:8080/
+  | draw                |                              |                   | repaint canvas, <br/> apply graphic changes to the shapes in the canvas                                                                                                                                                                         |
+  | add                 | Primitive shape              |                   | add shape to the canvas <br/> primitive shapes are available in the package "com.ait.lienzo.client.core.shape" <br/> e.g. var rect = new com.ait.lienzo.client.core.shape.Rectangle(50, 50);                                                    |
+  | getShape            | String id                    |     Primitive     | get shape instance                                                                                                                                                                                                                              |
+  | getWiresShape       | String id                    |    WiresShape     | get WiresShape instance <br/> wireshapes can group shapes and handle them as single composite "shape", besides the basic shape functionalities it adds connection capabilities throgh magnets and means to iterate through its grouped elements |
```
From the description, the usage is not clear. I would add similar description like there is one for `add()` function above.

handreyrc replied:
Same as above.

LuboTerifaj commented on packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/README.md:
```diff
@@ -73,24 +73,60 @@ Navigate to the test page in your root context path, eg: `http://localhost:8080/
+  | clearSelection         |                   |             | clear selection                                                                                                            |
+  | getSelectedNode        |                   |    Node     | get selected Node instance                                                                                                 |
+  | getSelectedEdge        |                   |    Edge     | get selected Edge instance                                                                                                 |
+  | getSelectedDefinition  |                   | Definition  | get selected Definition instance <br/> it is the data model of the selelected node with all declared properties and values |
```
```suggestion
  | getSelectedDefinition  |                   | Definition  | get selected Definition instance, <br/> it is the data model of the selelected node with all declared properties and values |
```

LuboTerifaj replied:
I would add more information here. It's not clear where the Definition can be found, so users can not find what properties they can use.

handreyrc replied:
Same as above.

LuboTerifaj reviewed (changes requested):
Hello @handreyrc ,

We are getting really close.

I reported two issues:

- [KOGITO-8964](https://issues.redhat.com/browse/KOGITO-8964) _Stunner - setBackroundColor function sets black color for non-existent color values_
- [KOGITO-8965](https://issues.redhat.com/browse/KOGITO-8965) _Stunner - Add function for getting Edge Ids to Session API_

We can make a call regarding:

- `center()` function
- Domain Definitions API
  - I see you removed it from the PR, but let's talk about it.

See my inline comments regarding documentation.

Thank you very much for the clarification regarding "Other thoughts" from my previous review.

handreyrc commented on packages/serverless-workflow-diagram-editor/sw-editor/sw-editor-kogito-app/README.md:
```diff
@@ -73,24 +73,60 @@ Navigate to the test page in your root context path, eg: `http://localhost:8080/
+  | getViewport         |                              |     Viewport      | get canvas Viewport instance                                                                                                                                                                                                                    |
+  | getNativeContext    |                              |  NativeContext2D  | get canvas NativeContext2D instance                                                                                                                                                                                                             |
+  | draw                |                              |                   | repaint canvas, <br/> apply graphic changes to the shapes in the canvas                                                                                                                                                                         |
+  | add                 | Primitive shape              |                   | add shape to the canvas <br/> primitive shapes are available in the package "com.ait.lienzo.client.core.shape" <br/> e.g. var rect = new com.ait.lienzo.client.core.shape.Rectangle(50, 50);                                                    |
```
Hey @LuboTerifaj @romartin ,

As we talked the API implemented in this PR will be exposed internally only. 
We still need to expose it from the envelope in a new Jira/PR, so developers can finally consume them. The documentation will also be added when we expose the API through the envelope so it can be automatically generated from the ts code. 
I'm also removing the codumentation from the Readme.md once it is not ne proper place to create and mantain it.

Thanks.

handreyrc commented:
Hey @LuboTerifaj and @romartin,

As agreed I updated the Jira, removed the documentation from the Readme.MD and rebased this PR.
 
I hope we are good to move forward with this PR.

Thanks for all your work and support here!
